File: src/test/java/com/github/squirrelgrip/meter/service/MeterReaderServiceTest.java
package com.github.squirrelgrip.meter.service;

import com.github.squirrelgrip.meter.database.DatabaseSession;
import com.github.squirrelgrip.meter.database.DatabaseSessionFactory;
import com.github.squirrelgrip.meter.domain.ContextKey;
import com.github.squirrelgrip.meter.exception.MeterReaderException;
import com.github.squirrelgrip.meter.exception.UnknownException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;

import java.io.BufferedReader;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;
import java.util.Map;

import static com.github.squirrelgrip.meter.domain.ContextKey.FILE_NAME;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;
import static org.mockito.Mockito.verifyNoInteractions;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
public class MeterReaderServiceTest {
    private MeterReaderService testSubject;

    @Mock
    private LineHandler lineHandler;
    @Mock
    private ErrorHandler errorHandler;

    @Mock
    private DatabaseSessionFactory databaseSessionFactory;
    @Mock
    private DatabaseSession databaseSession;
    @Mock
    private BufferedReader mockBufferedReader;

    @BeforeEach
    void beforeEach() {
        testSubject = new MeterReaderService(lineHandler, errorHandler, databaseSessionFactory);

        when(databaseSessionFactory.create()).thenReturn(databaseSession);
    }

    @Test
    void process() throws Exception {
        when(databaseSession.isTransactionActive()).thenReturn(false);
        InputStream inputStream = Thread.currentThread().getContextClassLoader().getResourceAsStream("sample.txt");

        Map<ContextKey, Object> context = new HashMap<>();
        context.put(FILE_NAME, "sample.txt");
        testSubject.process(inputStream, context);

        verify(lineHandler, times(14)).processLine(anyString(), eq(context));
        verify(databaseSession).close();
        verifyNoMoreInteractions(databaseSession);
    }

    @Test
    void process_Missing900Record() throws Exception {
        when(databaseSession.isTransactionActive()).thenReturn(true);
        InputStream inputStream = Thread.currentThread().getContextClassLoader().getResourceAsStream("missing900.txt");

        Map<ContextKey, Object> context = new HashMap<>();
        context.put(FILE_NAME, "sample.txt");
        testSubject.process(inputStream, context);

        verify(lineHandler, times(13)).processLine(anyString(), eq(context));
        verify(databaseSession).close();
        verify(databaseSession).isTransactionActive();
        verifyNoMoreInteractions(databaseSession);
    }

    @Test
    void process_GivenInputStreamIsEmpty() {
        ByteArrayInputStream inputStream = new ByteArrayInputStream("".getBytes());

        testSubject.process(inputStream, new HashMap<>());

        verifyNoInteractions(lineHandler);
    }

    @Test
    void process_GivenInputStreamContainsNoData() {
        ByteArrayInputStream inputStream = new ByteArrayInputStream("\n \n \n ".getBytes());

        testSubject.process(inputStream, new HashMap<>());

        verifyNoInteractions(lineHandler);
    }

    @Test
    void process_GivenReadingFromInputStreamThrowsException() throws IOException {
        IOException exception = new IOException();
        when(mockBufferedReader.readLine()).thenThrow(exception);

        HashMap<ContextKey, Object> metaData = new HashMap<>();
        testSubject.process(mockBufferedReader, metaData);

        verifyNoInteractions(lineHandler);
        verify(errorHandler).handleException(isA(UnknownException.class));
    }


}

File: src/test/kotlin/com/github/squirrelgrip/meter/service/MeterReaderServiceTest.kt
package com.github.squirrelgrip.meter.service

import com.github.squirrelgrip.meter.database.DatabaseSession
import com.github.squirrelgrip.meter.database.DatabaseSessionFactory
import com.github.squirrelgrip.meter.domain.MeterContext
import com.github.squirrelgrip.meter.entity.MeterReading
import com.github.squirrelgrip.meter.exception.UnknownException
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.junit.jupiter.MockitoExtension
import java.io.BufferedReader
import java.io.ByteArrayInputStream
import java.io.IOException

@ExtendWith(MockitoExtension::class)
class MeterReaderServiceTest {
    private lateinit var testSubject: MeterReaderService

    @Mock
    private lateinit var errorHandler: ErrorHandler

    @Mock
    private lateinit var databaseSessionFactory: DatabaseSessionFactory

    @Mock
    private lateinit var session: DatabaseSession

    @Mock
    private lateinit var mockBufferedReader: BufferedReader

    @BeforeEach
    fun beforeEach() {
        testSubject = MeterReaderService(errorHandler, databaseSessionFactory)

        `when`(databaseSessionFactory.create()).thenReturn(session)
    }

    @Test
    fun process() {
        `when`(session.isTransactionActive).thenReturn(false)
        val inputStream = Thread.currentThread().getContextClassLoader().getResourceAsStream("sample.txt")!!

        val context: MeterContext = testSubject.createMeterContext("sample.txt")
        testSubject.process(inputStream, context)

        verify(session).start()
        verify(session).close()
        verify(session).commit()
        verify(session, times(384)).save(any(MeterReading::class.java))
        verify(session).isTransactionActive
        verifyNoMoreInteractions(session)
    }

    @Test
    fun process_Missing900Record() {
        `when`(session.isTransactionActive).thenReturn(true)
        val inputStream = Thread.currentThread().getContextClassLoader().getResourceAsStream("missing900.txt")!!

        val context: MeterContext = testSubject.createMeterContext("sample.txt")
        testSubject.process(inputStream, context)

        verify(session).start()
        verify(session).close()
        verify(session).rollback()
        verify(session, times(384)).save(any(MeterReading::class.java))
        verify(session).isTransactionActive
        verifyNoMoreInteractions(session)
    }

    @Test
    fun process_GivenInputStreamIsEmpty() {
        val inputStream = ByteArrayInputStream("".toByteArray())

        testSubject.process(inputStream)
    }

    @Test
    fun process_GivenInputStreamContainsNoData() {
        val inputStream = ByteArrayInputStream("\n \n \n ".toByteArray())

        testSubject.process(inputStream)
    }

    @Test
    fun process_GivenReadingFromInputStreamThrowsException() {
        doThrow(IOException()).`when`(mockBufferedReader).readLine()

        testSubject.process(mockBufferedReader)

        verify(errorHandler).handleException(any(UnknownException::class.java))
    }
}
